### PR TITLE
ux-redesign: Redesign utilization disk card

### PR DESCRIPTION
**Pull request by bond95**
Fixes: https://github.com/oVirt/ovirt-web-ui/issues/778

Running VM with Guest Agent:
image
image

Running VM without Guest Agent or without OS at all
image

Stopped VM:
image

Running Windows:
image

### CODE REVIEW

**Comment by lwrigh** [comment]
This is looking great @bond95! Overall everything looks good to me, I have a few comments for the messaging though. 

For the no guest agent message, I would suggest having it say, "It looks like no guest agent is configured on the VM.". 

I would also suggest changing the no data available message to, "Utilization data is only available when the VM is running.".

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 1]
```diff
@@ -19,15 +18,45 @@ import style from './style.css'
...
+  return <div className={style['disk-fs-box']}>
+    <div className={style['disk-fs-name']}>{path}</div>
+    <div className={style['disk-fs-bar']}>
+      <UtilizationBar
+        now={usedInPercent}
+        thresholdWarning={thresholdWarning}
+        thresholdError={thresholdError}
+      />
+    </div>
+    <div className={style['disk-fs-used']}><strong>{round(value.used, 0)} of {round(value.total, 0)} {unit}</strong> Used</div>
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: The `strong` should be in the style instead ... `font-weight: bold;`

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 2]
```diff
@@ -43,18 +72,20 @@ const DiskCharts = ({ vm, isRunning, id, ...props }) => {
...
   const total = round(disks.provisionedSize, 1)
 
   return (
     <UtilizationCard className={style['chart-card']} id={id}>
       <CardTitle>Disk <span style={{ fontSize: '55%', verticalAlign: 'super' }}>(storage allocations)</span></CardTitle>
       <CardBody>
-        { !hasDisks &&
-          <NoLiveData id={`${id}-no-live-data`} message='This VM has no attached disks.' />
+        { !diskDetails && isRunning &&
+          <NoLiveData id={`${id}-no-live-data`} message='It seems that no guest agent is configurated on VM.' />
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: This reminds me that Utilization Card needs to be localized.  Perhaps you can start with this message.

File: src/ovirtapi/transform.js [hunk 3]
```diff
@@ -260,20 +260,24 @@ const VmStatistics = {
...
       // no values -> undefined, 1 value -> value.datum, >1 values -> [...values.datum]
+      // ?disks.usage -> {detail...}
       const datum =
         stat.values &&
         stat.values.value &&
-        (stat.values.value.length === 1
-          ? stat.values.value[0].datum
-          : stat.values.value.map(value => value.datum))
+        (stat.name === 'disks.usage'
+          ? JSON.parse(stat.values.value[0].detail || 'null')
```
Comment (sjd78) [inline src/ovirtapi/transform.js]: I recommend moving the JSON parsing down lower and have it `convertInt()` values here so the `parseInt` doesn't need to be done in the components.

````js
if (stat.name === 'disks.usage') {
  datum = JSON.parse(datum)
  datum.total = convertInt(datum.total)
  datum.used = convertInt(datum.used)
}
````

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 4]
```diff
@@ -43,18 +72,20 @@ const DiskCharts = ({ vm, isRunning, id, ...props }) => {
...
   const available = round(disks.provisionedSize - disks.actualSize, 1)
   const total = round(disks.provisionedSize, 1)
 
   return (
     <UtilizationCard className={style['chart-card']} id={id}>
       <CardTitle>Disk <span style={{ fontSize: '55%', verticalAlign: 'super' }}>(storage allocations)</span></CardTitle>
       <CardBody>
-        { !hasDisks &&
-          <NoLiveData id={`${id}-no-live-data`} message='This VM has no attached disks.' />
+        { !diskDetails && isRunning &&
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: Humm.... If the guest agent isn't reporting file system use, the disk allocated vs provisioned is still available to render.  I think it makes sense to put the `NoLiveData` messages below the `UtilizationCardDetails`.  The disk allocations, assuming the VM has attached disks, will always be available regardless of the state of the VM.

TL;DR - Put the "no attached disks" `NoLiveData` message check back here and move the 2 new `NoLiveData` messages to the same spot as the `DiskBar`s.

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 5]
```diff
@@ -43,18 +72,20 @@ const DiskCharts = ({ vm, isRunning, id, ...props }) => {
   })
 
   const { unit, value: disks } = convertValueMap('B', { actualSize, provisionedSize })
-  const used = round(disks.actualSize, 1)
   const available = round(disks.provisionedSize - disks.actualSize, 1)
   const total = round(disks.provisionedSize, 1)
 
   return (
     <UtilizationCard className={style['chart-card']} id={id}>
       <CardTitle>Disk <span style={{ fontSize: '55%', verticalAlign: 'super' }}>(storage allocations)</span></CardTitle>
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: We can probably take the super-scripted "storage allocations" off the `CardTitle` now.

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js (hunk 2 above)
Comment (bond95) [reply src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: It is better to do in separate PR, all at once: https://github.com/oVirt/ovirt-web-ui/issues/879

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js (hunk 1 above)
Comment (bond95) [reply src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: In this case I would prefer `strong`, because it will bold only some part of text, it is more readable, then adding a lot of `span`s and classes. Also this style would work for localization.

**Comment by bond95** [comment]
@lwrigh @sjd78 @gregsheremeta What do you think about this solution

With guest agent
image

Without guest agent
image

Without disks
image

VM is off
image

**Comment by lwrigh** [comment]
@bond95 It looks good to me!👍

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js (hunk 1 above)
Comment (sjd78) [reply src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: Good point.

**Comment by sjd78** [comment]
@bond95, @lwrigh, @gregsheremeta - When the VM is off, all of the attached disks and allocations are still known.  The card detail section can still be displayed (X unallocated of Y provisioned).

The chart below I think should only be displayed when guest agent data is available.

So, I see the variations as:
  1. No disks attached to the VM
  2. Disks attached, VM stopped (X unallocated of Y provisioned, no chart data)
  3. Disks attached, VM running, no guest agent (X of Y, no chart data)
  4. Disks attached, VM running, guest agent (X of Y, bar charts for file systems)

Looks like 1 and 4 are good.  2 and 3 are close, just need the X of Y headers.

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 6]
```diff
@@ -43,48 +74,60 @@ const DiskCharts = ({ vm, isRunning, id, ...props }) => {
...
 
   return (
     <UtilizationCard className={style['chart-card']} id={id}>
-      <CardTitle>Disk <span style={{ fontSize: '55%', verticalAlign: 'super' }}>(storage allocations)</span></CardTitle>
+      <CardTitle>Disk</CardTitle>
       <CardBody>
-        { !hasDisks &&
-          <NoLiveData id={`${id}-no-live-data`} message='This VM has no attached disks.' />
+        { !diskDetails && !isRunning &&
+          <NoLiveData id={`${id}-no-live-data`} />
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: I think this `NoLiveData` needs to go below the `UtilizationCardDetails`.

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 7]
```diff
@@ -43,48 +74,60 @@ const DiskCharts = ({ vm, isRunning, id, ...props }) => {
...
-            }}
-            tooltip={{
-              show: true,
-              contents: patternfly.pfDonutTooltipContents,
-            }}
-          />
-
-          {/* Disks don't have historic data */}
-        </React.Fragment>
+        { isRunning && vm.get('disks').size === 0 &&
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: `isRunning` is not necessary, doesn't matter if the VM is up or down, the "no disk is attached" message should still be displayed.

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 8]
```diff
@@ -43,48 +74,60 @@ const DiskCharts = ({ vm, isRunning, id, ...props }) => {
...
+          <NoLiveData id={`${id}-no-live-data`} message='It looks like no disk is attached to VM.' />
+        }
+        { isRunning && vm.get('disks').size > 0 &&
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: Here again, I don't think `isRunning` is necessary.  If disks are attached, the `UtilizationCardDetails` showing the allocated/provisioned disk total isn't dependent on the VM's state.

File: src/components/VmDetails/cards/UtilizationCard/DiskCharts.js [hunk 9]
```diff
@@ -43,48 +74,60 @@ const DiskCharts = ({ vm, isRunning, id, ...props }) => {
...
+          <React.Fragment>
+            <UtilizationCardDetails>
+              <UtilizationCardDetailsCount id={`${id}-available`}>{available}</UtilizationCardDetailsCount>
+              <UtilizationCardDetailsDesc>
+                <UtilizationCardDetailsLine1>Unallocated</UtilizationCardDetailsLine1>
+                <UtilizationCardDetailsLine2 id={`${id}-total`}>of {total} {unit} Provisioned</UtilizationCardDetailsLine2>
+              </UtilizationCardDetailsDesc>
+            </UtilizationCardDetails>
+            { !diskDetails &&
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/DiskCharts.js]: Are we sure that we want to keep the donut chart around for VMs without a guest agent?

File: src/components/VmDetails/cards/UtilizationCard/index.js [hunk 10]
```diff
@@ -45,8 +45,8 @@ const UtilizationCard = ({ vm }) => {
             }
           </Col>
           <Col>
-            { vm.has('disks')
-              ? <DiskCharts vm={vm} isRunning={isRunning} id={`${idPrefix}-disk`} />
+            { vm.has('disks') && stats.disks
```
Comment (sjd78) [inline src/components/VmDetails/cards/UtilizationCard/index.js]: If `stats.disks` is there, you won't be able to show the allocated/provisioned details for VMs without a guest agent.

Plus the checks in this component are really to guard against having the base VM data only and waiting for the VM's deep fetch to complete.

**Comment by bond95** [comment]
@lwrigh @sjd78 @gregsheremeta 

Running without Guest Agent
image

Running with Guest Agent
image

Not running
image

Without disks
image

**Comment by lwrigh** [comment]
Looks good to me!

**Comment by gregsheremeta** [comment]
@bond95 how easy is it to uppercase Windows drive letters?

`C:\`
`D:\`

vs.

`c:\`
`d:\`

**Comment by bond95** [comment]
ci build please

**Comment by bond95** [comment]
ci build please

**Comment by leistnerova** [comment]
ci build please

**Comment by gregsheremeta** [comment]
Honestly I think it looks pretty good and it's difficult to get perfect.

@lwrigh any thoughts about Lucie's comments above?

**Comment by lwrigh** [comment]
Ah I see, it's always a challenge to try and get text to behave nicely when the screen size gets compressed. I don't think scaling it down any more would be a good solution so I think for the time being wrapping is a better solution in this case.

**Comment by lwrigh** [comment]
@gregsheremeta @bond95 @sjd78 @leistnerova What do you guys think of these options for formatting? 
group 17
group 27

**Comment by bond95** [comment]
I like it.

**Comment by gregsheremeta** [comment]
Yep let's go with that. If we see issues in production we can try to enhance further.

**Comment by leistnerova** [comment]
I like it too, it makes higher only the row with long name and doesn't affect other mount points.

**Comment by lwrigh** [comment]
Great sounds good to me.

**Comment by bond95** [comment]
@lwrigh @leistnerova @gregsheremeta 

image

**Comment by bond95** [comment]
(rebased)

**Comment by lwrigh** [comment]
It looks good to me.

**Comment by leistnerova** [comment]
ci build please
